grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java

```diff
@@ -0,0 +1,366 @@
+                            e.setUnfinishedMessage(msg);
+                            throw e;
+                        }
+                        @SuppressWarnings("unchecked")  // We used the request marshaller so always safe.
+                        I castMsg = (I) msg;
+                        listener.onMessage(castMsg);
+                    } catch (InvalidProtocolBufferException e) {
+                        throw Status.INTERNAL.withDescription("Invalid protobuf byte sequence")
+                                             .withCause(e).asRuntimeException();
+                    }
+                }
+            } finally {
+                message.buf().release();
+            }
+        } else {
+            try (InputStream msg = message.stream()) {
+                if (isCancelled()) {
+                    return;
+                }
+                listener.onMessage(method.parseRequest(msg));
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    @Override
+    public void endOfStream() {
+        clientStreamClosed = true;
+        if (!closeCalled) {
+            listener.onHalfClose();
+        }
+    }
+
+    @Override
+    public void onError(Status status) {
+        close(status);
+    }
+
+    static HttpHeaders statusToTrailers(Status status, boolean headersSent) {
+        final HttpHeaders trailers;
+        if (headersSent) {
+            // Normal trailers.
+            trailers = new DefaultHttpHeaders();
+        } else {
+            // Trailers only response
+            trailers = new DefaultHttpHeaders(true, 3, true)
+                    .status(HttpStatus.OK)
+                    .set(HttpHeaderNames.CONTENT_TYPE, "application/grpc+proto");
+        }
+        trailers.add(GrpcHeaderNames.GRPC_STATUS, Integer.toString(status.getCode().value()));
+        if (status.getDescription() != null) {
+            trailers.add(GrpcHeaderNames.GRPC_MESSAGE,
+                         StatusMessageEscaper.escape(status.getDescription()));
+        }
+        return trailers;
+    }
+
+    HttpStreamReader messageReader() {
+        return messageReader;
+    }
+
+    void setListener(Listener<I> listener) {
+        checkState(this.listener == null, "listener already set");
+        this.listener = requireNonNull(listener, "listener");
+    }
+
+    private ByteBuf serializeMessage(O message) throws IOException {
+        if (message instanceof MessageLite) {
+            MessageLite msg = (MessageLite) message;
+            // Special case serializing a protobuf to a ByteBuffer, which is highly optimized.
+            ByteBuf buf = alloc.buffer(msg.getSerializedSize());
+            msg.writeTo(CodedOutputStream.newInstance(buf.nioBuffer(0, buf.writableBytes())));
+            buf.writerIndex(buf.capacity());
+            return buf;
+        }
+        CompositeByteBuf out = alloc.compositeBuffer();
```

Comment:
This, and the use of it in writeCompressed, is because a stream-copy usually involves an internal buffer that is written out many times. My assumption is CompositeByteBuffer will be faster for this case rather than a auto-growing ByteBuffer, but let me know if that's not true in your experience :)

Anyways, in practice no code will reach this for now since we only support protobuf, so it's hard to definitively say what the performance characteristic would be ;)